var/spack/repos/builtin/packages/gcc/package.py

```diff
@@ -264,6 +267,108 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
 
     build_directory = 'spack-build'
 
+    @property
+    def executables(self):
+        names = [r'gcc', r'[^\w]?g\+\+', r'gfortran']
+        suffixes = [r'-mp-\d\.\d', r'-\d\.\d', r'-\d', r'\d\d']
+        return [r''.join(x) for x in itertools.product(names, suffixes)]
+
+    @classmethod
+    def filter_detected_exes(cls, prefix, exes_in_prefix):
+        result = []
+        for exe in exes_in_prefix:
+            # clang++ matches g++ -> clan[g++]
+            if 'clang' in exe:
+                continue
+            # Filter out links in favor of real executables
+            if os.path.islink(exe):
+                continue
+            result.append(exe)
+        return result
+
+    @classmethod
+    def determine_version(cls, exe):
+        version_regex = re.compile(r'(.*)')
+        for vargs in ('-dumpfullversion', '-dumpversion'):
+            try:
+                output = spack.compiler.get_compiler_version_output(exe, vargs)
+                match = version_regex.search(output)
+                if match:
+                    return match.group(1)
+            except spack.util.executable.ProcessError:
+                pass
+            except Exception as e:
+                tty.debug(e)
+
+        return None
+
+    @classmethod
+    def determine_variants(cls, exes, version_str):
+        spec_str = 'gcc@{0} languages={1}'
+        languages, compilers = set(), {}
+        for exe in exes:
+            if 'gcc' in exe:
+                languages.add('c')
+                compilers['c'] = exe
+            if 'g++' in exe:
+                languages.add('c++')
+                compilers['cxx'] = exe
+            if 'gfortran' in exe:
+                languages.add('fortran')
+                compilers['fortran'] = exe
+
+        spec_str = spec_str.format(version_str, ','.join(languages))
+        return [Spec.from_detection(
+            spec_str, extra_attributes={'compilers': compilers}
+        )]
```

Comment:
We should modify this function so it only returns the variants, not the entire Spec

Reply:
It needs to return a list of variants + extra attributes. I'll try to think of some simpler interface.

Reply:
Maybe add an extra method for additional attributes? Can those attributes not be expressed in a spec string?

Reply:
> Maybe add an extra method for additional attributes? Can those attributes not be expressed in a spec string?

These attributes are precisely the ones that are not expressed in the spec string (e.g. the path to an external C compiler, the RPATHs it needs etc.). A new method would complicate package implementations and impose duplication, since when extra attributes are needed they are often computed using the same inspections as the variant.

That said, I completely see your point so I modified the API slightly in 5177578. Now `determine_variants` can either return:

1. A variant string (only variants like `languages=c,cxx,fortran`, not the entire spec)
2. A tuple of variant string + extra attributes
3. A list of items that fit either 1. or 2. (if more than one spec is found for the same prefix and version)

This should give enough flexibility to remain simple in most cases but be able to treat more complex cases when there's need for it. Does that seem better?